build_modules/lib/src/kernel_builder.dart

```diff
@@ -148,28 +143,68 @@ Future<void> _createKernel(
   }
 }
 
-/// Adds the source or kernel dependencies for [dependency] to
+/// Adds the source or kernel dependencies for [root] to
 /// [transitiveKernelDeps] or [transitiveSourceDeps].
 Future<void> _addModuleDeps(
-    Module dependency,
     Module root,
     List<AssetId> transitiveKernelDeps,
     List<AssetId> transitiveSourceDeps,
     BuildStep buildStep,
     String outputExtension) async {
-  var kernelId = dependency.primarySource.changeExtension(outputExtension);
-  if (await buildStep.canRead(kernelId)) {
-    // If we can read the kernel file, but it depends on any module in this
-    // package, then we need to only provide sources for that file since its
-    // dependencies in this package will only be providing sources as well.
-    if ((await dependency.computeTransitiveDependencies(buildStep))
-        .any((m) => m.primarySource.package == root.primarySource.package)) {
-      transitiveSourceDeps.addAll(dependency.sources);
+  final toCrawl = Queue.of([root.primarySource]);
+  final nodes = {root.primarySource: _ModuleNode(root, true)};
+
+  // Helper for any time we encounter a module which is provided via source and
+  // not kernel.
+  //
+  // For each of those we have to ensure that all parents of that module are
+  // provided via source as well.
+  void Function(AssetId) updateParents;
+  updateParents = (AssetId nodeId) {
```

Comment:
This doesn't look like it's reassigned so It doesn't need to be a variable.

```dart
void updateParents(AssetId nodeId) {
  ...
}
```

Reply:
Because it recursively calls itself it has to be, kind of annoying but :man_shrugging:

Reply:
nevermind I guess I am just crazy :man_facepalming: